Code/GraphMol/RGroupDecomposition/RGroupDecompData.h

```diff
@@ -383,21 +381,29 @@ struct RGroupDecompData {
 
   // compute the number of rgroups that would be added if we
   //  accepted this permutation
-  int compute_num_added_rgroups(std::vector<size_t> &tied_permutation) {
-    int num_added_rgroups = 0;
+  size_t compute_heavy_rgroup_counts(
+      const std::vector<size_t> &tied_permutation,
+      std::vector<int> &heavy_counts) {
+    size_t i = 0;
+    size_t num_added_rgroups = 0;
+    heavy_counts.resize(labels.size(), 0);
     for (int label : labels) {
-      if (label <= 0) {  // label is not user supplied
-        for (size_t m = 0; m < tied_permutation.size();
-             ++m) {  // for each molecule
-          auto rg = matches[m][tied_permutation[m]].rgroups.find(label);
-          if (rg != matches[m][tied_permutation[m]].rgroups.end()) {
-            if (!rg->second->is_hydrogen) {
-              num_added_rgroups += 1;  //= label;
-              break;
-            }
+      int incr = (label > 0) ? -1 : 1;
+      // int incr = 1;
+      bool incremented = false;
+      for (size_t m = 0; m < tied_permutation.size();
+           ++m) {  // for each molecule
+        auto rg = matches[m][tied_permutation[m]].rgroups.find(label);
+        if (rg != matches[m][tied_permutation[m]].rgroups.end() &&
+            !rg->second->is_hydrogen) {
+          if (label <= 0 && !incremented) {
+            incremented = true;
+            ++num_added_rgroups;
           }
+          heavy_counts[i] += incr;
```

Comment:
This doesn't make sense to me.

if the label is < 0 this just means that the group is not user defined, but here you are subtracting 1 from the heavy counts.  I don't think this is right unless I'm missing something.